MoreLinq/Duplicates.cs

```diff
@@ -0,0 +1,61 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2023 Julien Aspirot. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+
+    static partial class MoreEnumerable
+    {
+        /// <summary>
+        ///   Returns all duplicated elements of the given source.
+        /// </summary>
+        /// <param name="source">source sequence.</param>
+        /// <typeparam name="T">The type of the elements in the source sequence.</typeparam>
+        /// <returns>all elements that are duplicated.</returns>
+        public static IEnumerable<T> Duplicates<T>(this IEnumerable<T> source) => Duplicates(source, null);
+
+        /// <summary>
+        ///   Returns all duplicated elements of the given source, using the specified element equality comparer
+        /// </summary>
+        /// <param name="source">source sequence.</param>
+        /// <param name="comparer">The equality comparer to use to determine whether or not keys are equal.
+        /// If null, the default equality comparer for <c>TSource</c> is used.</param>
+        /// <typeparam name="TSource">Type of the source sequence</typeparam>
+        /// <returns>all elements of the source sequence that are duplicated, based on the provided equality comparer</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="source"/> is null.</exception>
+        public static IEnumerable<TSource> Duplicates<TSource>(this IEnumerable<TSource> source, IEqualityComparer<TSource>? comparer)
+        {
+            if (source is null) throw new ArgumentNullException(nameof(source));
+
+            return GetDuplicates();
+
+            IEnumerable<TSource> GetDuplicates()
+            {
+                var keySet = new HashSet<TSource>(comparer);
+                var yieldSet = new HashSet<TSource>(comparer);
+                foreach (var element in source)
+                {
+                    if (!keySet.Add(element) && yieldSet.Add(element))
+                    {
+                        yield return element;
+                    }
+                }
+            }
```

Comment:
@atifaziz  earlier today i did a benchmark using Dictionnary as data source vs my 2 hashsets:

![image](https://github.com/morelinq/MoreLINQ/assets/14810960/5c506eb4-4915-4aa9-832d-965ac4ecfdbc)


since ScanBy also use dictionnary do you still want me to use that implementation?

Reply:
I trust you've done it right, but do you mind sharing the benchmark code, even as [gist](https://gist.github.com/)? It would also help for future maintainers/visitors to have as a reference.

Do you also get the same numbers with .NET 8?

Reply:
<script src="https://gist.github.com/julienasp/663c30d1b11755652d1cd4bc08d70563.js"></script>

i did not try with .NET 8, it was run on .NET 6 but it probably better with .NET 8.

Also the benchmark was quick n dirty i did not try to use any worse case or best case scenario. My data generations was very plain.

Reply:
Thanks for sharing the benchmark code!

I think there's a fundamental flaw in there! In the `DuplicatesWithDictionary` version, [you're creating the `counts` dictionary _inside the loop_](https://gist.github.com/julienasp/663c30d1b11755652d1cd4bc08d70563#file-gistfile1-txt-L96), so _for each iteration_!

I've [published a revised version](https://gist.github.com/atifaziz/b72d4239145428638910aad8a19edb9c) that has 3 changes:

- The `counts` dictionary is created outside the loop
- There's a global setup where the source list is created to remove the invariant noise
- The results are simply consumed in for-loop as opposed to creating an array, again to remove invariant noise

I also tested on .NET 8 (though that's less important for now due to the fundamental flaw) and the results are drastically different:

```
BenchmarkDotNet v0.13.10, Windows 11 (10.0.22621.2715/22H2/2022Update/SunValley2)
Intel Core i7-1065G7 CPU 1.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.100
  [Host]     : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
```
| Method                      | Mean     | Error   | StdDev  | Gen0    | Gen1   | Allocated |
|---------------------------- |---------:|--------:|--------:|--------:|-------:|----------:|
| Duplicates_With_Two_HashSet | 213.2 μs | 2.48 μs | 2.32 μs | 27.8320 | 4.6387 | 114.71 KB |
| Duplicates_With_Dictionary  | 221.2 μs | 3.89 μs | 3.64 μs | 17.3340 | 1.9531 |  71.59 KB |

The `Duplicates_With_Dictionary` version is just about as fast and uses less memory. 😅

Reply:
omg that's embarassing 😥, i did that way too quicky good thing i shared the benchmark code lol

and i would have found the bug if tested correctly sorry 

@viceroypenguin  FYI, i fail hard on that quick benchmark

Reply:
do you want me to change the code to use the dictionnary version or the one with the scanby usage?

Reply:
![image](https://github.com/morelinq/MoreLINQ/assets/14810960/27bdec4d-95a2-4d17-a61f-3815c432a24b)


i pushed the one with scanby

Reply:
If we're willing to have different code for different TFMs, we can improve performance even further. The difference is that the new version only accesses the dictionary once and uses `ref` access to manipulate the value. This reduces the duplication of work in accessing the value twice. This version would need to be gated behind `#if NET6_0_OR_GREATER`.

```cs
IEnumerable<TSource> GetDuplicates()
{
	var counts = new Dictionary<TSource, int>(comparer);
	foreach (var element in source)
	{
		int CountElement()
		{
			ref var count = ref CollectionsMarshal.GetValueRefOrAddDefault(
				counts,
				element,
				out var _);
			return ++count;
		}

		if (CountElement() == 2)
			yield return element;
	}
}
```

```
// * Summary *

BenchmarkDotNet v0.13.10, Windows 11 (10.0.22621.2715/22H2/2022Update/SunValley2)
12th Gen Intel Core i7-12700H, 1 CPU, 20 logical and 14 physical cores
.NET SDK 8.0.100
  [Host] : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
```

| Method                      | Mean      | Error    | StdDev   | Gen0   | Gen1   | Allocated |
|---------------------------- |----------:|---------:|---------:|-------:|-------:|----------:|
| Duplicates_With_Two_HashSet | 102.13 μs | 1.047 μs | 0.875 μs | 9.2773 | 1.9531 | 114.71 KB |
| Duplicates_With_Dictionary  | 111.46 μs | 0.228 μs | 0.202 μs | 5.7373 | 0.7324 |  71.59 KB |
| Duplicates_With_Dictionary2 |  70.01 μs | 0.313 μs | 0.277 μs | 5.7373 | 0.7324 |  71.58 KB |

[code](https://gist.github.com/viceroypenguin/8a6ad84a77a3bc3979af0abf58fcd832)

Reply:
I don't think there's any need for the `CollectionsMarshal.GetValueRefOrAddDefault` sorcery or conditionally compiling separate versions. 😅 If you look at my original proposal with `ScanBy`, you don't really need to keep counting so you can stop updating after the first duplicate like so:

```c#
var counts = new Dictionary<TSource, int>(comparer);
foreach (var element in source)
{
    if (counts.TryGetValue(element, out var count))
    {
        if (count == 1)
        {
            yield return element;
            counts[element] = 2;
        }
    }
    else
    {
        counts[element] = 1;
    }
}
```

This version is just about as fast while being simple for anyone to understand. I [revised the benchmark](https://gist.github.com/atifaziz/b72d4239145428638910aad8a19edb9c/b18df8ec202b5273ae1ae876b90faa9cbc3caed7) to include the above and tested across .NET 6, 7, and 8:

```
BenchmarkDotNet v0.13.10, Windows 11 (10.0.22621.2715/22H2/2022Update/SunValley2)
Intel Core i7-1065G7 CPU 1.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.100
  [Host]   : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
  .NET 6.0 : .NET 6.0.25 (6.0.2523.51912), X64 RyuJIT AVX2
  .NET 7.0 : .NET 7.0.14 (7.0.1423.51910), X64 RyuJIT AVX2
  .NET 8.0 : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
```

| Method                      | Runtime  | Mean     | Error   | StdDev  | Ratio | Gen0    | Gen1   | Allocated | Alloc Ratio |
|---------------------------- |--------- |---------:|--------:|--------:|------:|--------:|-------:|----------:|------------:|
| Duplicates_With_Two_HashSet | .NET 6.0 | 226.6 μs | 3.44 μs | 3.05 μs |  1.00 | 27.8320 | 4.3945 | 114.71 KB |        1.00 |
| Duplicates_With_Dictionary  | .NET 6.0 | 227.7 μs | 3.22 μs | 3.01 μs |  1.00 | 17.3340 |      - |  71.59 KB |        0.62 |
| Duplicates_Via_ScanBy       | .NET 6.0 | 425.0 μs | 5.77 μs | 5.40 μs |  1.87 | 17.0898 | 2.4414 |  71.75 KB |        0.63 |
| Duplicates_Using_Ref_Count  | .NET 6.0 | 161.9 μs | 2.86 μs | 2.68 μs |  0.72 | 17.3340 |      - |  71.58 KB |        0.62 |
| Duplicates_With_Opt_Count   | .NET 6.0 | 164.8 μs | 2.58 μs | 2.29 μs |  0.73 | 17.3340 |      - |  71.58 KB |        0.62 |
|                             |          |          |         |         |       |         |        |           |             |
| Duplicates_With_Two_HashSet | .NET 7.0 | 227.5 μs | 2.84 μs | 2.66 μs |  1.00 | 27.8320 | 4.6387 | 114.71 KB |        1.00 |
| Duplicates_With_Dictionary  | .NET 7.0 | 234.6 μs | 3.48 μs | 3.26 μs |  1.03 | 17.3340 |      - |  71.59 KB |        0.62 |
| Duplicates_Via_ScanBy       | .NET 7.0 | 440.3 μs | 4.05 μs | 3.59 μs |  1.94 | 17.0898 | 2.4414 |  71.75 KB |        0.63 |
| Duplicates_Using_Ref_Count  | .NET 7.0 | 178.9 μs | 2.60 μs | 2.03 μs |  0.79 | 17.3340 |      - |  71.58 KB |        0.62 |
| Duplicates_With_Opt_Count   | .NET 7.0 | 180.2 μs | 3.40 μs | 5.95 μs |  0.79 | 17.3340 | 0.2441 |  71.58 KB |        0.62 |
|                             |          |          |         |         |       |         |        |           |             |
| Duplicates_With_Two_HashSet | .NET 8.0 | 214.7 μs | 4.22 μs | 6.69 μs |  1.00 | 27.8320 | 4.6387 | 114.71 KB |        1.00 |
| Duplicates_With_Dictionary  | .NET 8.0 | 223.0 μs | 3.10 μs | 2.75 μs |  1.04 | 17.3340 | 1.9531 |  71.59 KB |        0.62 |
| Duplicates_Via_ScanBy       | .NET 8.0 | 278.7 μs | 5.44 μs | 5.34 μs |  1.30 | 17.0898 | 2.4414 |  71.75 KB |        0.63 |
| Duplicates_Using_Ref_Count  | .NET 8.0 | 138.6 μs | 2.60 μs | 2.56 μs |  0.65 | 17.3340 | 2.1973 |  71.58 KB |        0.62 |
| Duplicates_With_Opt_Count   | .NET 8.0 | 146.3 μs | 2.81 μs | 2.49 μs |  0.68 | 17.3340 | 2.1973 |  71.58 KB |        0.62 |

As you can see, it's super fast across all .NET versions and with same memory profile. The `ScanBy` version is about _twice_ as slow on .NET 6 and 7 than the two hash sets version from @julienasp, but only 30% slower on .NET 8, which is remarkable. I guess it just comes down more calls and indirections just to lambdas. All the versions using dictionary will also need to be adapted for `null` keys since `HashSet<>` permits it, but `Dictionary<,>` imposes the `notnull` constraint on the key.

Also bear in mind that while these benchmarks are interesting, ultimately _actual performance_ is what counts and will vary a lot with data (uniqueness of keys, hash collisions, etc.) and the application's environment like the runtime, JIT, tiered compilation, PGO optimisations, AOT and what not in the future. The key takeaway should be: make it work, make it right and then make it fast (if necessary). The last bit could have been left for another PR. As I said before, LINQ is generally not going to be used on hot paths so simplicity, abstraction, reuse and composition should come first.

Reply:
> reuse and composition should come first.

so we keep the ScanBy implementation?

Reply:
> so we keep the ScanBy implementation?

Think so, for now and for sake of getting this in. We can always improve it later.

BTW, just to complete the performance discussion and demonstrate how it can vary with _data_, if you change [the benchmark setup](https://gist.github.com/atifaziz/b72d4239145428638910aad8a19edb9c/b18df8ec202b5273ae1ae876b90faa9cbc3caed7#file-program-cs-L17-L24) so that all keys are unique:

```diff
diff --git a/Program.cs b/Program.cs
index 498e36f..dead2a8 100644
--- a/Program.cs
+++ b/Program.cs
@@ -19,7 +19,7 @@ public class Benchmarks
     {
         for (var i = 0; i < 15000; i++)
         {
-            list.Add(i % 1000);
+            list.Add(i);
         }
     }
 
```

the your version with two hash sets is the fatest _and_ takes less memory:

```
BenchmarkDotNet v0.13.10, Windows 11 (10.0.22621.2715/22H2/2022Update/SunValley2)
Intel Core i7-1065G7 CPU 1.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.100
  [Host]   : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
  .NET 6.0 : .NET 6.0.25 (6.0.2523.51912), X64 RyuJIT AVX2
  .NET 7.0 : .NET 7.0.14 (7.0.1423.51910), X64 RyuJIT AVX2
  .NET 8.0 : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
```

| Method                      | Runtime  | Mean     | Error    | StdDev   | Ratio | RatioSD | Gen0     | Gen1     | Gen2     | Allocated | Alloc Ratio |
|---------------------------- |--------- |---------:|---------:|---------:|------:|--------:|---------:|---------:|---------:|----------:|------------:|
| Duplicates_With_Two_HashSet | .NET 6.0 | 320.0 μs |  3.79 μs |  3.36 μs |  1.00 |    0.00 |  95.2148 |  95.2148 |  95.2148 | 526.23 KB |        1.00 |
| Duplicates_With_Dictionary  | .NET 6.0 | 477.9 μs |  7.96 μs |  7.05 μs |  1.49 |    0.02 | 124.0234 | 124.0234 | 124.0234 | 657.46 KB |        1.25 |
| Duplicates_Via_ScanBy       | .NET 6.0 | 654.7 μs | 11.64 μs | 14.71 μs |  2.05 |    0.05 | 124.0234 | 124.0234 | 124.0234 | 657.63 KB |        1.25 |
| Duplicates_Using_Ref_Count  | .NET 6.0 | 424.8 μs |  4.06 μs |  3.60 μs |  1.33 |    0.02 | 124.5117 | 124.5117 | 124.5117 | 657.46 KB |        1.25 |
| Duplicates_With_Opt_Count   | .NET 6.0 | 485.5 μs |  9.44 μs |  9.27 μs |  1.52 |    0.03 | 124.5117 | 124.5117 | 124.5117 | 657.46 KB |        1.25 |
|                             |          |          |          |          |       |         |          |          |          |           |             |
| Duplicates_With_Two_HashSet | .NET 7.0 | 327.3 μs |  5.65 μs |  5.29 μs |  1.00 |    0.00 |  95.2148 |  95.2148 |  95.2148 | 526.23 KB |        1.00 |
| Duplicates_With_Dictionary  | .NET 7.0 | 482.6 μs |  6.30 μs |  5.89 μs |  1.47 |    0.03 | 124.5117 | 124.5117 | 124.5117 | 657.46 KB |        1.25 |
| Duplicates_Via_ScanBy       | .NET 7.0 | 703.2 μs | 13.89 μs | 14.27 μs |  2.15 |    0.07 | 124.0234 | 124.0234 | 124.0234 | 657.63 KB |        1.25 |
| Duplicates_Using_Ref_Count  | .NET 7.0 | 435.3 μs |  6.18 μs |  5.78 μs |  1.33 |    0.03 | 124.5117 | 124.5117 | 124.5117 | 657.46 KB |        1.25 |
| Duplicates_With_Opt_Count   | .NET 7.0 | 497.1 μs |  9.10 μs |  9.74 μs |  1.52 |    0.04 | 124.0234 | 124.0234 | 124.0234 | 657.46 KB |        1.25 |
|                             |          |          |          |          |       |         |          |          |          |           |             |
| Duplicates_With_Two_HashSet | .NET 8.0 | 315.9 μs |  6.25 μs |  5.85 μs |  1.00 |    0.00 |  95.2148 |  95.2148 |  95.2148 | 526.23 KB |        1.00 |
| Duplicates_With_Dictionary  | .NET 8.0 | 496.7 μs |  8.78 μs | 13.13 μs |  1.59 |    0.06 | 124.0234 | 124.0234 | 124.0234 | 657.46 KB |        1.25 |
| Duplicates_Via_ScanBy       | .NET 8.0 | 573.2 μs | 11.62 μs | 33.53 μs |  1.91 |    0.12 | 124.0234 | 124.0234 | 124.0234 | 657.63 KB |        1.25 |
| Duplicates_Using_Ref_Count  | .NET 8.0 | 417.4 μs |  7.92 μs |  7.02 μs |  1.32 |    0.04 | 124.5117 | 124.5117 | 124.5117 | 657.46 KB |        1.25 |
| Duplicates_With_Opt_Count   | .NET 8.0 | 480.2 μs |  8.15 μs |  7.63 μs |  1.52 |    0.04 | 124.5117 | 124.5117 | 124.5117 | 657.46 KB |        1.25 |

So what case are you optmising for? That there'll be many duplicates or there'll be few to none? Unless you go through the trouble of something adaptive at run-time, the actual mileage will vary.

Reply:
> So what case are you optmising for? That there'll be many duplicates or there'll be few to none?

in the specifics usecases where i used it, and why i wanted a HasDuplicates it was for invariants validations. So few to none.

but im good with the scan by as a first iteration!